### PR TITLE
Add retry logic around the websocket connection

### PR DESCRIPTION
**Pull request by zippolyte**
*Note: Please remember to review the Datadog [Contribution Guidelines](https://github.com/DataDog/dd-agent/blob/master/CONTRIBUTING.md)
if you have not yet done so.*

### What does this PR do?

Add a retry mechanism when the web socket connection closes, to try and reopen a new one instead of crashing the nozzle

### Motivation

When under high loads, the web socket connection can close, and that made the nozzle crash while we could instead reopen the connection.

### Testing Guidelines

An overview on [testing](https://github.com/DataDog/dd-agent/blob/master/tests/README.md)
is available in our contribution guidelines.

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: datadogfirehosenozzle/datadog_firehose_nozzle.go [hunk 1]
```diff
@@ -103,16 +104,35 @@ func (d *DatadogFirehoseNozzle) Start() error {
...
+		d.startWorkers()
+		// Start infinite loop to periodically submit metrics (metricsMap) to Datadog
+		err = d.postToDatadog()
+		if !shouldReconnect(err) {
+			break
+		}
+
+		// memorize the retry timestamp
+		retryHistory = append(retryHistory[1:], time.Now())
+		if retryHistory[2].Sub(retryHistory[0]) < 5*time.Second {
```
Comment (gmmeyer) [inline datadogfirehosenozzle/datadog_firehose_nozzle.go]: what if retryHistory[2] doesn't exist? Will this error?

Comment (zippolyte) [reply datadogfirehosenozzle/datadog_firehose_nozzle.go]: I initialize the slice with 3 elements, and the logic makes it that there are always 3 elements

File: datadogfirehosenozzle/datadog_firehose_nozzle.go [hunk 2]
```diff
@@ -103,16 +104,38 @@ func (d *DatadogFirehoseNozzle) Start() error {
...
+		d.startWorkers()
+		// Start infinite loop to periodically submit metrics (metricsMap) to Datadog
+		err = d.postToDatadog()
+
+		if !shouldReconnect(err) {
+			break
+		}
+
+		// memorize the retry timestamp
+		retryHistory = append(retryHistory[1:], time.Now())
```
Comment (gmmeyer) [inline datadogfirehosenozzle/datadog_firehose_nozzle.go]: this is nice

File: datadogfirehosenozzle/datadog_firehose_nozzle_test.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package datadogfirehosenozzle_test
+package datadogfirehosenozzle
```
Comment (gmmeyer) [inline datadogfirehosenozzle/datadog_firehose_nozzle_test.go]: 🍰

File: datadogfirehosenozzle/datadog_firehose_nozzle_test.go [hunk 4]
```diff
@@ -199,6 +199,53 @@ var _ = Describe("Datadog Firehose Nozzle", func() {
...
+					Value: proto.Float64(float64(i)),
+					Unit:  proto.String("gauge"),
+				},
+				Deployment: proto.String("deployment-name"),
+				Job:        proto.String("doppler"),
+			}
+			fakeFirehose.AddEvent(envelope)
+		}
+
+		fakeFirehose.SetCloseMessage(websocket.FormatCloseMessage(websocket.ClosePolicyViolation, "Client did not respond to ping before keep-alive timeout expired."))
```
Comment (gmmeyer) [inline datadogfirehosenozzle/datadog_firehose_nozzle_test.go]: 👍

**Pull request merged**